### PR TITLE
Retry on watch longpoll disconnect.

### PR DESCRIPTION
**Pull request by mitchdenny**
Partially addresses #2422 

This PR introduces a Polly resilience pipeline to retry the watch resource long poll operation when it disconnects. We detect the disconnect by overving that the exception that is thrown has an inner exception of type `EndOfStreamException`. If we see this then we'll log a debug message for future supportability purposes, but otherwise exit.

We swallow `OperationCancelledExceptions` and exit cleanly. All other exceptions are logged.
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/3029)

### CODE REVIEW

File: src/Aspire.Hosting/Dcp/ApplicationExecutor.cs [hunk 1]
```diff
@@ -156,21 +158,44 @@ private void WatchResourceChanges(CancellationToken cancellationToken)
...
+                        }
+                        finally
+                        {
+                            semaphore.Release();
+                        }
                     }
-                }
+                }, cancellationToken).ConfigureAwait(false);
             }
             catch (Exception ex) when (ex is not OperationCanceledException)
```
Comment (mitchdenny) [inline src/Aspire.Hosting/Dcp/ApplicationExecutor.cs:200]: We preserve existing behavior where we swallow all exceptions and log an error except for cancellations which bubble out.

File: src/Aspire.Hosting/Dcp/ApplicationExecutor.cs [hunk 2]
```diff
@@ -156,21 +158,44 @@ private void WatchResourceChanges(CancellationToken cancellationToken)
...
-                await foreach (var (eventType, resource) in kubernetesService.WatchAsync<T>(cancellationToken: cancellationToken))
+                ShouldHandle = new PredicateBuilder().HandleInner<EndOfStreamException>(),
+                BackoffType = DelayBackoffType.Constant,
+                MaxRetryAttempts = int.MaxValue,
+                UseJitter = true,
+                MaxDelay = TimeSpan.FromSeconds(30),
+                OnRetry = (retry) =>
                 {
-                    await semaphore.WaitAsync(cancellationToken).ConfigureAwait(false);
+                    _logger.LogDebug(
```
Comment (mitchdenny) [inline src/Aspire.Hosting/Dcp/ApplicationExecutor.cs:170]: This is hard to test. You have to drastically alter KubernetesService to inject this error synthetically and DCP has quite a long timeout for long poll (like Kubernetes). I've asked @karolz-ms for an environment variable we can inject into DCP when launching it to bring down the long poll timeout for resource watching so we can simulate this more easily.

**Comment by davidfowl** [comment]
cc @martincostello

File: src/Aspire.Hosting/Dcp/ApplicationExecutor.cs [hunk 3]
```diff
@@ -156,21 +158,44 @@ private void WatchResourceChanges(CancellationToken cancellationToken)
...
-                    try
-                    {
-                        await handler(eventType, resource).ConfigureAwait(false);
-                    }
-                    finally
+            var pipeline = new ResiliencePipelineBuilder().AddRetry(retryUntilCancelled).Build();
+
+            try
+            {
+                await pipeline.ExecuteAsync(async (pipelineCancellationToken) =>
```
Comment (martincostello) [inline src/Aspire.Hosting/Dcp/ApplicationExecutor.cs:183]: Is it much refactoring/cognitive overload to refactor this to pass the state through into Polly so you can avoid the need to allocation the closure here by making the lambda static, or is the context where this happens not perf-sensitive enough for that to be a concern?

Comment (mitchdenny) [reply src/Aspire.Hosting/Dcp/ApplicationExecutor.cs:183]: Its not that perf sensitive. We execute this once per resource type of project, executable, and container. And if we fail with an end of stream (after ~30 mins) its not a huge cost.

File: src/Aspire.Hosting/Dcp/ApplicationExecutor.cs [hunk 4]
```diff
@@ -185,25 +187,51 @@ private void WatchResourceChanges(CancellationToken cancellationToken)
 
         async Task WatchKubernetesResource<T>(Func<WatchEventType, T, Task> handler) where T : CustomResource
         {
+            var retryUntilCancelled = new RetryStrategyOptions()
+            {
+                ShouldHandle = new PredicateBuilder().HandleInner<EndOfStreamException>(),
+                BackoffType = DelayBackoffType.Exponential,
+                MaxRetryAttempts = int.MaxValue,
+                UseJitter = true,
+                MaxDelay = TimeSpan.FromSeconds(30),
```
Comment (davidfowl) [inline src/Aspire.Hosting/Dcp/ApplicationExecutor.cs:196]: 30 seconds seems short no?

**Review by davidfowl — APPROVED** [review]

File: src/Aspire.Hosting/Dcp/ApplicationExecutor.cs [hunk 5]
```diff
@@ -185,25 +187,51 @@ private void WatchResourceChanges(CancellationToken cancellationToken)
...
+                            semaphore.Release();
+                        }
                     }
-                }
+                }, cancellationToken).ConfigureAwait(false);
             }
             catch (Exception ex) when (ex is not OperationCanceledException)
             {
-                _logger.LogError(ex, "Watch task over kubernetes {ResourceType} resources terminated", typeof(T).Name);
+                _logger.LogCritical(ex, "Watch task over kubernetes {ResourceType} resources terminated", typeof(T).Name);
```
Comment (davidfowl) [inline src/Aspire.Hosting/Dcp/ApplicationExecutor.cs:234]: Can we say something more about what action the user can take. Like mmake sure dcp is running etc./

File: src/Aspire.Hosting/Dcp/ApplicationExecutor.cs (hunk 4 above)
Comment (mitchdenny) [reply src/Aspire.Hosting/Dcp/ApplicationExecutor.cs:196]: Discussed offline. This is the only code that is hitting the dcp resource watch endpoint, hitting it every 30 seconds is probably fine. We hit it more frequently than that with an exponential backoff to 30 seconds. Realistically - if it doesn't respond by 30 seconds it is probably dead and a different (fatal) exception would have thrown and we would have aborted.

**Pull request merged**